### PR TITLE
Fix x.pow(0) gradient when x contains 0

### LINKED ISSUE #8940: Issue in a particular case in backpropagation
The back propagation isn't working properly in the case 0**0. For example in the code below we have an "nan" which should mathematically not be here.

x = torch.tensor([0., 1.], requires_grad=True)
loss = sum((x)**0)
loss.backward()
print(loss)
print(x.grad)

tensor(2.)
tensor([nan.,   0.])

It is probably because the derivative of loss is 0*(x^-1) which is 0/0 is the case x=0. The gradient should be 0. In fact the formula (x^n)' = n*x^(n-1) which is probably used, is false for n=0.

This is a special case but can happen.

Hope it can help

Hv0nnus

### PR DESCRIPTION
**Pull request by vishwakftw**
This closes https://github.com/pytorch/pytorch/issues/8940 .

### CODE REVIEW

File: tools/autograd/templates/Functions.cpp [hunk 1]
```diff
@@ -108,6 +108,24 @@ Tensor norm_backward(Tensor grad, const Tensor & self, const Scalar & p_, Tensor
...
+  if (exponent == 0.0) {
+    return zeros_like(self);
+  } else {
+    return grad * exponent * self.pow(exponent - 1);
+  }
+}
+
+Tensor pow_backward_self(Tensor grad, const Tensor & self, const Tensor & exponent) {
+  Tensor zero_mask = (exponent == 0.0);
+  return at::where(zero_mask, zeros_like(self), grad * exponent * self.pow(exponent - 1));
```
Comment (vadimkantorov) [inline tools/autograd/templates/Functions.cpp]: This line and the one above makes two allocations of the size of the input tensor. Are there any perf implications?

Comment (vishwakftw) [reply tools/autograd/templates/Functions.cpp]: Thanks for catching this. There can be one fix:
```
return at::where((exponent == 0.0), at::zeros({1}, grad.type()), grad * exponent * self.pow(exponent - 1));
```

I don't think

**Review by ssnl — DISMISSED** [review]
need test in test_autograd.py

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
@vadimkantorov Unfortunately not. This might be a good addition (and pretty easy to add), if there are other cases where we might need it. Maybe we should generalize this into some sort of arbitrary equality test against a floating point number? (CC @colesbury @cpuhrsch for more opinions).

I do have a question beyond what this patch does: when the exponent is close to zero (but not exactly zero), what happens to the gradients? Is it numerically stable? If it's not, it would be nice (though not strictly necessary) to fix that too.

**Comment by vishwakftw** [comment]
@ezyang As the exponent tends to 0, the derivative is 0 for non-zero `x` tends to `inf` for `x` = 0.

**Comment by vadimkantorov** [comment]
@ezyang Thinking more about it, merging a comparison binary op and `torch.where` goes along the op fusion line (maybe it's a good case for the automatic fuser?). Comparison + torch.where is a frequent use-case I guess, but surfacing all comparison ops through `where` may bring unjustified api complexity.

**Comment by vishwakftw** [comment]
I think for the purposes of this PR, the usage of `where` can remain as designed at the moment. Probably, when the semantics of `where` change, this part of the code can be revisited and modified accordingly. What do you think @ezyang ?

**Comment by ezyang** [comment]
Works for me.
